### PR TITLE
refactor: allocate table id in the procedure

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. refactor: replace TableMetadataManager with TableNameManager
2. refactor: allocate table id in the procedure
3. feat: add test utils for testing CreateTableProcedure
4. test: add tests for CreateTableProcedure

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
#3233

### CODE REVIEW

File: src/common/meta/src/ddl/utils.rs [hunk 1]
```diff
@@ -27,19 +27,17 @@ use crate::key::TableMetadataManagerRef;
 use crate::peer::Peer;
 use crate::rpc::ddl::CreateTableTask;
 
-pub fn handle_operate_region_error(datanode: Peer) -> impl FnOnce(Error) -> Error {
+/// Adds [Peer] context if the error is unretryable.
+pub fn add_peer_context_if_need(datanode: Peer) -> impl FnOnce(Error) -> Error {
```
Comment (fengjiachun) [inline src/common/meta/src/ddl/utils.rs:31]: Perhaps "add_peer_context_if_needed" or "add_peer_context_as_needed" would be more suitable names.

File: src/common/procedure/src/procedure.rs [hunk 2]
```diff
@@ -57,6 +57,18 @@ impl Status {
...
 
+    /// Downcasts [Status::Done]'s output to &T
+    ///  #Panic:
+    /// - if [Status] is not the [Status::Done].
+    /// - if the output is None.
+    pub fn downcast_output_ref<T: 'static>(&self) -> Option<&T> {
+        if let Status::Done { output } = self {
+            output.as_ref().unwrap().downcast_ref()
+        } else {
+            unreachable!("")
```
Comment (fengjiachun) [inline src/common/procedure/src/procedure.rs:68]: Let's replace empty strings with some useful information

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/procedure/src/procedure.rs [hunk 3]
```diff
@@ -57,6 +57,21 @@ impl Status {
         Status::Done { output: None }
     }
 
+    /// Downcasts [Status::Done]'s output to &T
+    ///  #Panic:
+    /// - if [Status] is not the [Status::Done].
+    /// - if the output is None.
+    pub fn downcast_output_ref<T: 'static>(&self) -> Option<&T> {
```
Comment (MichaelScofield) [inline src/common/procedure/src/procedure.rs:64]: Seems this method is only used in tests, better move there.

File: src/common/meta/src/ddl/tests/create_table.rs
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/ddl/tests/create_table.rs:1]: Did you mean to rename the dir to `test`?

File: src/common/meta/src/ddl/create_table.rs [hunk 4]
```diff
@@ -124,6 +139,22 @@ impl CreateTableProcedure {
...
+            .table_metadata_allocator
+            .create(
+                &TableMetadataAllocatorContext {
+                    cluster_id: self.creator.data.cluster_id,
+                },
+                &self.creator.data.task,
+            )
+            .await?;
+        self.creator
+            .set_allocated_metadata(table_id, table_route, region_wal_options);
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/ddl/create_table.rs:157]: If the procedure allocates metadata without persisting, then we will lose the previously allocated metadata forever.

File: src/common/meta/src/ddl/tests/create_table.rs
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/ddl/tests/create_table.rs:1]: > Did you mean to rename the dir to `test`?

If that I will refactor this in #3293

File: src/common/meta/src/ddl/create_table.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/common/meta/src/ddl/create_table.rs:157]: I mean we should check whether the metadata has been allocated before create it. Like this:
```rust
if self.creator.allocated_metadata().is_none() {
  // create metadata
  ...

  // set metadata
  self.creator.set_allocated_metadata(...);
}
```

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**